python/hail/methods/statgen.py

```diff
@@ -2908,15 +2909,104 @@ def filter(self):
+        raise ValueError("memory per core must be greater than {} MB".format(min_memory_per_core * 1024 * 1024))
+    bytes_per_variant = math.ceil(8 * n_samples / genotypes_per_pack) + variant_byte_overhead
+    memory_available_per_core = bytes_per_core * fraction_memory_to_use
+    max_queue_size = int(max(1.0, math.ceil(memory_available_per_core / bytes_per_variant)))
+
+    sites_only_table = Table(Env.hail().methods.LocalLDPrune.apply(
+        require_biallelic(ds, 'ld_prune')._jvds, float(r2), window, max_queue_size))
+
+    sites_path = new_temp_file()
+    sites_only_table.write(sites_path, overwrite=True)
+    sites_only_table = hl.read_table(sites_path)
+
+    locally_pruned_ds = ds.annotate_rows(pruned=sites_only_table[ds.row_key])
+    locally_pruned_ds = locally_pruned_ds.filter_rows(hl.is_defined(locally_pruned_ds.pruned))
+
+    locally_pruned_ds = (locally_pruned_ds.annotate_rows(
+        mean=locally_pruned_ds.pruned.mean, 
+        centered_length_sd_reciprocal=locally_pruned_ds.pruned.centered_length_sd_reciprocal)
+        .add_row_index('row_idx'))
+
+    locally_pruned_path = new_temp_file()
+    locally_pruned_ds.write(locally_pruned_path, overwrite=True)
+    locally_pruned_ds = hl.read_matrix_table(locally_pruned_path)
+
+    normalized_mean_imputed_genotype_expr = (
+        hl.cond(hl.is_defined(locally_pruned_ds.GT),
+                (locally_pruned_ds.GT.n_alt_alleles() - locally_pruned_ds.mean)
+                * locally_pruned_ds.centered_length_sd_reciprocal, 0))
+
+    # BlockMatrix.from_entry_expr writes to disk
+    block_matrix = BlockMatrix.from_entry_expr(normalized_mean_imputed_genotype_expr, block_size=1024)
+    correlation_matrix = (block_matrix @ (block_matrix.T))
+
+    entries = Table(correlation_matrix._jbm.filteredEntriesTable(
+        locally_pruned_ds.annotate_rows(contig=locally_pruned_ds.locus.contig, pos=locally_pruned_ds.locus.position)
+            .select_rows('contig', 'pos').rows().key_by("contig")._jt, window, False
+    ))
+    entries_path = new_temp_file()
+    entries.write(entries_path, overwrite=True)
+    entries = hl.read_table(entries_path)
+
+    entries = entries.filter((entries.entry ** 2 >= r2) & (entries.i != entries.j) & (entries.i < entries.j))
+
+    index_table = sites_only_table.add_index().select('locus', 'idx').key_by('idx')
+    entries = entries.annotate(locus_i=index_table[entries.i].locus, locus_j=index_table[entries.j].locus)
+
+    entries = entries.filter((entries.locus_i.contig == entries.locus_j.contig)
+                             & ((hl.abs(entries.locus_i.position - entries.locus_j.position)) <= window))
+
+    related_nodes_to_remove = maximal_independent_set(entries.i, entries.j, keep=False)
+
+    pruned_ds = locally_pruned_ds.filter_rows(
+        hl.is_defined(related_nodes_to_remove[locally_pruned_ds.row_idx]), keep=False)
+
+    info("LD prune step 2 of 2: nVariantsKept={}".format(pruned_ds.count_rows()))
```

Comment:
So the old code called cache so counts were free, but now count is quite expensive here (pruned_ds is will be computed twice).  I say just get rid of this.